python/src/poses_bindings.cpp

```diff
@@ -374,7 +477,37 @@ void export_poses()
     {
         MAKE_PTR(CPose3DPDFGaussian)
 
+        // dealing with overloading
+        void (CPose3DPDF::*getCovariance1)(mrpt::math::CMatrixFixedNumeric<double,6ul,6ul> &cov) const = &CPose3DPDFGaussian::getCovariance;
+
         class_<CPose3DPDFGaussian, bases<CPose3DPDF> >("CPose3DPDFGaussian", init<>())
+            .def(init<CPose3DPDFGaussian>())
+            .def(init<CPose3D>())
+            .def("getMean", &CPose3DPDFGaussian::getMean, "Returns the mean, or mathematical expectation of the probability density distribution (PDF).")
+            .def("getCovariance", getCovariance1, "Returns an estimate of the pose covariance matrix (STATE_LENxSTATE_LEN cov matrix).")
+            .def("getCovarianceAndMean", &CPose3DPDFGaussian::getCovarianceAndMean, "Returns an estimate of the pose covariance matrix (STATE_LENxSTATE_LEN cov matrix) and the mean, both at once.")
+            .def("saveToTextFile", &CPose3DPDFGaussian::saveToTextFile, "Save PDF's particles to a text file. See derived classes for more information about the format of generated files.")
+            .def("changeCoordinatesReference", &CPose3DPDF::changeCoordinatesReference, "this = p (+) this. This can be used to convert a PDF from local coordinates to global, providing the point (newReferenceBase) from which 'to project' the current pdf. Result PDF substituted the currently stored one in the object.")
+            .def("getCovarianceEntropy", &CPose3DPDFGaussian::getCovarianceEntropy, "Compute the entropy of the estimated covariance matrix.")
+            .def("drawSingleSample", &CPose3DPDFGaussian::drawSingleSample, "Draws a single sample from the distribution")
+//            .def("copyFrom", &CPose3DPDFGaussian::copyFrom, "Copy operator, translating if necesary (for example, between particles and gaussian representations).")
+            .def("bayesianFusion", &CPose3DPDF::bayesianFusion, "Bayesian fusion of two pose distributions (product of two distributions->new distribution), then save the result in this object (WARNING: See implementing classes to see classes that can and cannot be mixtured!).")
+            .def("inverse", &CPose3DPDFGaussian::inverse, "Returns a new PDF such as: NEW_PDF = (0,0,0) - THIS_PDF")
+            .def("jacobiansPoseComposition", &CPose3DPDFGaussian::jacobiansPoseComposition, "This static method computes the pose composition Jacobians.")
+            .def("__str__", &CPose3DPDFGaussian_asString)
+            .def_readwrite("mean", &CPose3DPDFGaussian::mean)
+            // Note: the following does not work unfortunately: p.cov[35] = 1.0
+            // FIXME: change cov to return actual matrix object (use def_readwrite)
+            .add_property("cov", CPose3DPDFGaussian_get_cov, CPose3DPDFGaussian_set_cov)
```

Comment:
Hi guys!

Checked the diff, and I think it is ok! But I have a problem now... when I
was writing the bindings a year ago, I had to stop instantly, but it seems
that I did some changes which are not uploaded to github yet :-( .  The
changes are related to object inheritance ( I also had a look how you (Nik)
did it ;-) ) :  instead of using MAKE_PTR macro I switched to MAKE_PTR_BASE
macro to setup object inheritance.

To solve it I will update my master branch when all other changes are
included, backup MY (not uploaded) python-bindings branch, and manually
merge changes to the then up-to-date master. Is my solution acceptable?

@ Nik: I'm following you on Google+ now... ;-) (me: Peter Rudolph,
semael23@gmail.com)

Cheers,
Peter (sem23)


2016-03-22 20:44 GMT+01:00 Nikolaus Demmel <notifications@github.com>:

> In python/src/poses_bindings.cpp
> <https://github.com/MRPT/mrpt/pull/209#discussion_r57055478>:
>
> > +            .def("getMean", &CPose3DPDFGaussian::getMean, "Returns the mean, or mathematical expectation of the probability density distribution (PDF).")
> > +            .def("getCovariance", getCovariance1, "Returns an estimate of the pose covariance matrix (STATE_LENxSTATE_LEN cov matrix).")
> > +            .def("getCovarianceAndMean", &CPose3DPDFGaussian::getCovarianceAndMean, "Returns an estimate of the pose covariance matrix (STATE_LENxSTATE_LEN cov matrix) and the mean, both at once.")
> > +            .def("saveToTextFile", &CPose3DPDFGaussian::saveToTextFile, "Save PDF's particles to a text file. See derived classes for more information about the format of generated files.")
> > +            .def("changeCoordinatesReference", &CPose3DPDF::changeCoordinatesReference, "this = p (+) this. This can be used to convert a PDF from local coordinates to global, providing the point (newReferenceBase) from which 'to project' the current pdf. Result PDF substituted the currently stored one in the object.")
> > +            .def("getCovarianceEntropy", &CPose3DPDFGaussian::getCovarianceEntropy, "Compute the entropy of the estimated covariance matrix.")
> > +            .def("drawSingleSample", &CPose3DPDFGaussian::drawSingleSample, "Draws a single sample from the distribution")
> > +//            .def("copyFrom", &CPose3DPDFGaussian::copyFrom, "Copy operator, translating if necesary (for example, between particles and gaussian representations).")
> > +            .def("bayesianFusion", &CPose3DPDF::bayesianFusion, "Bayesian fusion of two pose distributions (product of two distributions->new distribution), then save the result in this object (WARNING: See implementing classes to see classes that can and cannot be mixtured!).")
> > +            .def("inverse", &CPose3DPDFGaussian::inverse, "Returns a new PDF such as: NEW_PDF = (0,0,0) - THIS_PDF")
> > +            .def("jacobiansPoseComposition", &CPose3DPDFGaussian::jacobiansPoseComposition, "This static method computes the pose composition Jacobians.")
> > +            .def("__str__", &CPose3DPDFGaussian_asString)
> > +            .def_readwrite("mean", &CPose3DPDFGaussian::mean)
> > +            // Note: the following does not work unfortunately: p.cov[35] = 1.0
> > +            // FIXME: change cov to return actual matrix object (use def_readwrite)
> > +            .add_property("cov", CPose3DPDFGaussian_get_cov, CPose3DPDFGaussian_set_cov)
>
> Before people start using this, I would also like to get this FIXME
> addressed. I don't have to time to look into this right now though.
> Basically, the idea would be to provide proper member access, such that
> setting individual array elements works like expected, but at the same time
> still provide a convenient way to convert to/from list or possibly numpy
> array (see the TODO entry in math_bindings.cpp added also in this PR).
>
> @sem23 <https://github.com/sem23>, any opinion on that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/MRPT/mrpt/pull/209/files/e71a54b6eea45fb8a5026e0e02cf62938c612538#r57055478>
>